### ISSUE TITLE
OSX OpenGL compatibility

### ISSUE THREAD

**Issue by twinaphex** [libretro/beetle-psx-libretro#140]
To @simias - Here is the current compilation log -

http://hastebin.com/johogubicu.vbs

OSX is stuck at GL 4.1-ish or so I believe, so we'll have to make it more backwards compatible.

**Comment by simias** [comment]
`glMapBufferRange` is in OpenGL 3.0 so I thought that'd be low enough for Mac as well. I guess I could use multiple buffers instead of splitting a single big one but it's more cumbersome that way. Might as well drop VAOs completely if I do that I guess, that'll be one fewer gl3 dependency.

I couldn't find a lot of reference to this problem online, only [this page](https://developer.apple.com/library/content/releasenotes/General/MacOSXLionAPIDiffs/OpenGL.html) which seems to hint that the define was added in Mac OS 10.7 but I don't really know what to make of it.

Maybe as a workaround we could use `GL_MAP_INVALIDATE_BUFFER_BIT` or maybe even nothing at all, although that might cause some performance issues if it stalls everything everytime we remap the buffer.

**Comment by twinaphex** [comment]
We could make mednafen psx HW OSX 10.7-only for now, although we still want to have code paths later on so that we can go as high or low with GL support as potentially possible.

**Comment by simias** [comment]
I'm not sure I understand you, do you mean that you were testing with an OSX version older than 10.7?

If I understand correctly "Mac OS X 10.7 Lion" was released in October 2012 and given Apple's notorious planned obsolescence policy that seems pretty ancient.

**Comment by Paradise12314** [comment]
Sorry to hijack this post, but it's sort of relevant. What version of OpenGL is needed for Windows build? My GPU (9500gt) is OGL 3.3. 

No matter what, Retroarch crashes when selecting a game. I have the correct Bios files with  correct MD5 checksum, and all my Cue files are correct. Only issue I can imagine, is your Rewrite that bumped OGL to 4.4 some months back.

**Comment by saftle** [comment]
@Paradise12314 yes the OGL renderer was bumped up, and they are working on ways of making it backwards compatible.

**Comment by micochops** [comment]
Hello!

Bumping this.
Having a difficult time making the Mednafen PSX HW core work.
The software core, I have no problems with.

May I know what version of OpenGL is supported by this core?
The reason I ask is because I'm currently using a 13" Macbook Pro (mid-2012).
It has a Core i5-3210M processor and Intel HD Graphics 4000 GPU which I believe only supports up to [OpenGL v 4.1 on OSX](https://en.wikipedia.org/wiki/List_of_Intel_graphics_processing_units#Seventh_generation).

Thanks!

**Comment by benneq** [comment]
Still no news?

**Comment by rz5** [comment]
@micochops @benneq: @simias downgraded it to about OpenGL 3.2. So if you can get it to compile on macOS, you should be able to use the GL renderer.

**Comment by twinaphex** [comment]
Well, there are still some GL functions used which are higher than that. The downgraded version of that function caused issues with the graphics so I had to revert it back to the GL 4.x function.

**Comment by twinaphex** [comment]
In specific, this is the function I'm talking about - define NEW_COPY_RECT in order to test it -

https://github.com/libretro/beetle-psx-libretro/blob/master/rsx/rsx_lib_gl.cpp#L2666

**Comment by benneq** [comment]
Can this be done using OpenGL 4.1? Doesn't seem that macOS user will get more than 4.1 even with High Sierra

**Comment by micochops** [comment]
Me too, I believe OpenGL 4.1 will be the highest macOS will receive.

**Comment by externo6** [comment]
Has this been parked or still worked on?

**Comment by twinaphex** [comment]
@externo6 Zero activity I am afraid. I would still very much like to see this resolved but we are lacking in OpenGL coders able to take this on (and having enough willingness to do so, and also having access to a Mac).

**Comment by externo6** [comment]
@twinaphex Thats a shame but understandable. Thanks for the update :)
